### ISSUE TITLE
global declarations made available for all files as soon as the associated declared module is imported

### ISSUE THREAD

**Issue by kingatlas** [Microsoft/TypeScript#16396]
**TypeScript Version:**  2.3.4

**Code**

```ts
// first.ts
const obs : KnockoutObservable<string> = ko.observable("hello"); // should raise error TS2304: Cannot find name 'ko'. and error TS2304: Cannot find name 'KnockoutObservable'.
console.log(obs());

// second.ts
import * as ko from "knockout"; // this import makes 'ko' globally available even for first.ts

export class Foo {
    bar =  ko.observable("bar");
}

// tsconfig.json
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "es5",
        "noImplicitAny": true,
        "sourceMap": false,
        "declaration": false,
        "noUnusedLocals": true,
        "outDir": "dist",
        "types": [
        ]
    },
    "exclude": [
        "node_modules",
        "dist"
    ]
}

```

**Expected behavior:**
We should get compilation errors for first.ts
error TS2304: Cannot find name 'ko'
error TS2304: Cannot find name 'KnockoutObservable'

**Actual behavior:**
Compilation succeed as soon as we import "knockout" somewhere.
see this repo for the overall solution: https://github.com/kingatlas/typescript-sandbox

**Comment by mhegazy** [comment]
A global scope mutation is noticed throughout the system once it happens. and the compiler cannot detect order of execution. it is possible that you have your module at runtime executing before your other file, this is knowledge that is external to the compilation.
Ideally you should not have any global scope polluters like that. one possible fix is to remove the global mutations from knockout.

**Comment by kingatlas** [comment]
Thank you for the answer,
Can we prevent executing global declarations when importing a module using "import from ..." ?

**Comment by mhegazy** [comment]
> Can we prevent executing global declarations when importing a module using "import from ..." ?

no, since modules can still put/augment declarations in the global scope.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
